common/lib/new.cc

```diff
@@ -0,0 +1,369 @@
+          last_block_node->m_next = block_node_ptr->m_next;
+          if (last_block_node->m_next) {
+            last_block_node->m_next->m_prev = last_block_node;
+          }
+          m_used -= sizeof(block_node);
+          m_free += sizeof(block_node);
+          continue;
+        }
+      }
+
+      last_block = &block;
+    }
+  }
+
+  void set_new_limit(std::size_t new_limit) {
+    if (new_limit >= (m_heap_limit + MIN_ALLOC_SIZE)) {
+      add_block(
+          // new block address will be directly after the region defined by the
+          // previous limit.
+          &__heap_start + m_heap_limit, new_limit - m_heap_limit);
+
+      m_heap_limit = new_limit;
+    }
+  }
+
+private:
+  // nodes are added at the end of the list.
+  void add_node(block_node *node) {
+    list_node *node_last = m_head;
+    for (; node_last->m_next; node_last = node_last->m_next) {
+    }
+
+    node_last->m_next = node;
+    node->m_prev = node_last;
+  }
+
+  void add_block(std::byte *address, std::size_t size) {
+    // assume zero alignment padding.
+    add_node(new (address) block_node{size});
+    m_used += sizeof(block_node);
+    m_free += m_head->m_block.m_size;
+  }
+
+  block_node *m_head = nullptr;
+
+public:
+  // heap limit is essentially a global var.  It can
+  // be set prior to ctor of blocklist.
+  static constexpr std::size_t HEAP_DEFAULT_LIMIT = 4096;
+  static std::size_t m_heap_limit;
+
+  // accounting.
+  std::size_t m_free = 0;
+  std::size_t m_used = 0;
+};
+
+blocklist &get_free_list() noexcept {
+  static blocklist free_list;
+  return free_list;
+}
+
+void run_new_handler() {
+  const auto newp = std::get_new_handler();
+  if (newp) {
+    newp();
+  }
+}
+} // namespace
+
+std::size_t blocklist::m_heap_limit = SIZE_MAX;
+
+void *operator new(std::size_t size) {
+  const auto block = operator new(size, std::nothrow);
+  if (!block) {
+    run_new_handler();
+  }
+
+  return block;
+}
+
+void *operator new[](std::size_t size) { return operator new(size); }
+
+void *operator new(std::size_t count, const std::nothrow_t &) noexcept {
+  auto &free_list = get_free_list();
+  return free_list.split_block(free_list.find_first_fit(count), count);
+}
+
+void *operator new[](std::size_t count, const std::nothrow_t &) noexcept {
+  return operator new(count, std::nothrow);
+}
+
+void *operator new(std::size_t, void *ptr) { return ptr; }
+
+void *operator new[](std::size_t, void *ptr) { return ptr; }
+
+void operator delete(void *ptr) noexcept {
+  if (!ptr) {
+    return;
+  }
+
+  auto &free_list = get_free_list();
+  free_list.free_block(block::get_block(static_cast<std::byte *>(ptr)));
+  free_list.coallesce_blocks();
+}
+
+void operator delete[](void *ptr) noexcept { operator delete(ptr); }
+
+// helper functions for getting/setting flags in guard_object
+static bool initializerHasRun(std::uint8_t *guard_object) {
+  return (*guard_object != 0);
+}
+
+static void setInitializerHasRun(std::uint8_t *guard_object) {
```

Comment:
As I was scrolling around the llvm source, I found that this was copied from the first Apple Google search result, comments and all.

The Apache license requires that this sort of thing include the original copyright notice in the source file. Please either do this, or rewrite the copied portions in your own terms.

There's probably a fair use argument here, but such arguments have become increasingly fraught as of late, and it's easier and safer to just follow the rules.